flink-connectors/flink-connector-opensearch/src/main/java/org/apache/flink/streaming/connectors/opensearch/sink/OpenSearchWriter.java

```diff
@@ -0,0 +1,334 @@
+                                    .setHttpClientConfigCallback(
+                                            httpClientBuilder ->
+                                                    httpClientBuilder.setDefaultCredentialsProvider(
+                                                            credentialsProvider)));
+            // TODO: Other auth implementations (sigv4, etc..)
+        } else {
+            client = new RestHighLevelClient(RestClient.builder(HttpHost.create(osUrl)));
+        }
+        return client;
+    }
+
+    @Override
+    public void write(InputT element, Context context) throws IOException, InterruptedException {
+        // do not allow new bulk writes until all actions are flushed
+        while (checkpointInProgress) {
+            mailboxExecutor.yield();
+        }
+        emitter.emit(element, context, requestIndexer);
+    }
+
+    @Override
+    public List prepareCommit(boolean flush) throws IOException, InterruptedException {
+        checkpointInProgress = true;
+        while (pendingActions != 0 && (flushOnCheckpoint || flush)) {
+            bulkProcessor.flush();
+            LOG.info("Waiting for the response of {} pending actions.", pendingActions);
+            mailboxExecutor.yield();
+        }
+        checkpointInProgress = false;
+        return Collections.emptyList();
+    }
+
+    @Override
+    public void close() throws Exception {
+        closed = true;
+        emitter.close();
+        bulkProcessor.close();
+        client.close();
+    }
+
+    private class DefaultRequestIndexer implements RequestIndexer {
+
+        @Override
+        public void add(IndexRequest... indexRequests) {
+            for (final IndexRequest indexRequest : indexRequests) {
+                pendingActions++;
+                bulkProcessor.add(indexRequest);
+            }
+        }
+
+        @Override
+        public void add(UpdateRequest... updateRequests) {
+            for (final UpdateRequest updateRequest : updateRequests) {
+                pendingActions++;
+                bulkProcessor.add(updateRequest);
+            }
+        }
+
+        @Override
+        public void add(DeleteRequest... deleteRequests) {
+            for (final DeleteRequest deleteRequest : deleteRequests) {
+                pendingActions++;
+                bulkProcessor.add(deleteRequest);
+            }
+        }
+    }
+
+    private void enqueueActionInMailbox(
+            ThrowingRunnable<? extends Exception> action, String actionName) {
+        // If the writer is cancelled before the last bulk response (i.e. no flush on checkpoint
+        // configured or shutdown without a final
+        // checkpoint) the mailbox might already be shutdown, so we should not enqueue any
+        // actions.
+        if (isClosed()) {
+            return;
+        }
+        mailboxExecutor.execute(action, actionName);
+    }
+
+    private void extractFailures(BulkRequest request, BulkResponse response) {
```

Comment:
does this do anything to know which failures are "retryable"?

Reply:
As of now, It doesn't classify the failures. There will be a retry for all failures. But it honors the bulkBackoffMaxRetries which is set to 3 by default.